### ISSUE TITLE
File Analysis Hanging on what seems to be numpy

### ISSUE THREAD

**Issue by QuantFreedom1022** [microsoft/pylance-release#4062]
here is a loom video of me explaining the problem
https://www.loom.com/share/75a5ced878f64390b24ad6e0d80f381b

please watch the video because it is to hard to explain but i will try

What seems to be happening is in the latest version 2023.3.20 all the way up until 2023.1.20 the file analyzer seems to be hanging on this one situation with numpy. which means if I make any changes I'm not able to see any of my problems or errors or anything like I showed in the video but if I go to version 2023.1.10 everything seems to be working perfectly fine.

 again please watch the video and look at the logs if you understand the logs in order to get a better understanding because this is just my guess based on the output it seems to hang on numpy and it is hanging on a numpy array that is trying to be called into a function

 I have included the code as a paste bin link and also I have included logs from the version that works and I have included logs from one version that doesn't work. although I have tried all versions from 2023.1.20 all the way up until the latest and none of them work

 if anybody needs any more information please let me know because I tried to do my best to explain what's going on and hopefully it makes enough sense

<h3> Environment data </h3>

-   Language Server version: 2023.3.20
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: off

<h3> Code Snippet </h3>

Python code causing the issue https://pastebin.com/qw2dvJ3z

<h3> Logs </h3>

[Log from the version that works 2023.1.10](https://github.com/microsoft/pylance-release/files/10949550/good.log)
[log from version that doesn't work 2023.3.20](https://github.com/microsoft/pylance-release/files/10949553/bad.log)

<h3> Pip List </h3>
https://pastebin.com/vW7N5kHB

**Comment by bschnurr** [comment]
Thank you for the detailed write up. investigating now

**Comment by bschnurr** [comment]
i'm having issues finding the following imports

```
from vectorbt.qf.portfolio.enums.enums_qf import *
from vectorbt.qf.portfolio.nb.execute_funcs import *
```

missing code for `to_1d_array_nb_qf`

setting extra perf trace settings
```
"python.analysis.logTypeEvaluationTime": true,
"python.analysis.typeEvaluationTimeThreshold": 500,
```

we appear to be spending a lot of time figuring out the type of these variables.

```
(20124) [BG(1)] getType ["max_equity_risk_value_array" (simple) [90:5]] (1833ms) [f:24, t:54, p:120, i:551, b:214]
(20124) [BG(1)] getType ["size_pct_array" (simple) [93:5]] (971ms)
(20124) [BG(1)] getType ["sl_pcts_array" (simple) [100:5]] (563ms)
(20124) [BG(1)] getType ["sl_to_be_when_pct_from_avg_ent <shortened> " (simple) [116:5]] (534ms)
(20124) [BG(1)] getType ["tsl_pcts_array" (simple) [123:5]] (607ms)
(20124) [BG(1)] getType ["tsl_trail_by_array" (simple) [132:5]] (544ms)
(20124) [BG(1)] getType ["tsl_when_pct_from_avg_entry_ar <shortened> " (simple) [138:5]] (632ms)
(20124) [BG(1)] getType ["tp_pcts_array" (simple) [145:5]] (548ms)
(20124) [BG(1)] getType ["arrays" (simple) [150:5]] (854ms)
```

can you send me the logs with those 2 settings enabled for the broken case?
do you know how I can resolve the missing code for `to_1d_array_nb_qf` ?

**Comment by bschnurr** [comment]
wild guess by maybe its due to all type annotations you have using `tp.ArrayLike`, which is a union type of a whole bunch of things.  I'll try removing them or just adding the annotation for the type you are assigning it to.. ie. float or bool 

```
  fee_pct: float = 0.06,
    leverage: tp.ArrayLike = 0,
    max_equity_risk_pct: tp.ArrayLike = np.nan,
    max_equity_risk_value: tp.ArrayLike = np.nan,
    max_lev: float = 100.,
    max_order_size_pct: float = 100.,
    max_order_size_value: float = np.inf,
    min_order_size_pct: float = .01,
    min_order_size_value: float = 1.,
    mmr: float = 0.5,
    size_pct: tp.ArrayLike = np.nan,
    size_value: tp.ArrayLike = np.nan,
    slippage_pct: float = np.nan,
 
    # Stop Losses
    sl_pcts: tp.ArrayLike = np.nan,
    sl_prices: tp.ArrayLike = np.nan,
 
    sl_to_be: tp.ArrayLike = False,
    sl_to_be_based_on: tp.ArrayLike = np.nan,
    sl_to_be_then_trail: tp.ArrayLike = False,
    sl_to_be_trail_when_pct_from_avg_entry: tp.ArrayLike = np.nan,
    sl_to_be_when_pct_from_avg_entry: tp.ArrayLike = np.nan,
    sl_to_be_zero_or_entry: tp.ArrayLike = np.nan,
 
    # Trailing Stop Loss Params
    tsl_pcts = np.nan,
    tsl_prices: tp.ArrayLike = np.nan,
 
    tsl_based_on: tp.ArrayLike = np.nan,
    tsl_true_or_false: tp.ArrayLike = False,
    tsl_trail_by: tp.ArrayLike = np.nan,
    tsl_when_pct_from_avg_entry: tp.ArrayLike = np.nan,
 
    # Take Profit Params
    risk_rewards: tp.ArrayLike = np.nan,
    tp_pcts: tp.ArrayLike = np.nan,
    tp_prices: tp.ArrayLike = np.nan,
 
```

**Comment by QuantFreedom1022** [comment]
@bschnurr yeah it could be that but the problem is that I'm kind of doing the same thing all over the place and it just happens to be with one very tiny instance and it worked perfectly fine until that one exact situation that I'm talking about. Because it works super fast and works perfectly fine all over the place except for that one specific instance and it has to be commented out so that's the part that is really tricky. But like I said I'm reworking the whole code and so after I'm done if it's still happening I'll let you know but feel free to try whatever you need to try. If you need any of the code Source you can just go to my GitHub and go to the Quant freedom repository and all of the Imports and stuff are all there

**Comment by bschnurr** [comment]
thanks. i'll try the whole repo

**Comment by QuantFreedom1022** [comment]
Sorry I forgot to mention that it's on the development Branch it's not on the main branch

**Comment by bschnurr** [comment]
@erictraut maybe a good stress test of complexity.

Python code causing the issue https://pastebin.com/qw2dvJ3z

pip install numpy
pip install numba
pip install vectorbt


You might need to tweak the max complexity to look at union types and possibility unknown function count.


the boolean `moved_sl_to_be_order_result` gets set to `False` before a forloop but then get re-assigned to unknown due to a function call returning tuple.

some of the code is missing, I can't find `to_1d_array_nb_qf()` in the dev branch either.

**Comment by QuantFreedom1022** [comment]
@bschnurr sorry i took out all the _qf in my dev branch ... if you go to [the helper functions ](https://github.com/QuantFreedom1022/QuantFreedom/blob/dev/quantfreedom/backtester/nb/helper_funcs.py) you will see the to_1d array

again all of this is on the dev branch

**Comment by erictraut** [comment]
@QuantFreedom1022, thanks for sharing the video. If and when you have the code in a state where it reproduces the hang that you were seeing earlier, please let us know. Until then, I don't think there's much we can do.

**Comment by bschnurr** [comment]
Not seeing any perf issue when opening `quantfreedom/backtester/nb/simulate.py`

the new annotations of `PossibleArray` might have improved this.

please reopen if you find something again.

**Comment by QuantFreedom1022** [comment]
@bschnurr yeah maybe ... ever since your suggestion i went through and tried to make all my code as efficient as possible ... to the degree of knowledge that i have of python ... i also just realized that i am on the latest version of pylance and dealing with no issues at all ... so thanks again for the suggestion because maybe that was the thing that did it
